### ISSUE TITLE
[Perf Guide] Flame graph scripts on Asterinas

### ISSUE THREAD

**Issue by junyang-zh** [asterinas/asterinas#964]
[Flame graph](https://github.com/brendangregg/FlameGraph) is a well known and powerful tool for performance (bottoleneck) analysis. It's based on sampling. If you inspect the call stack 100 times per second, the function that appears more often, would like to consume more time. The flame graph helps you to visualize it.

Here's my experience about how to capture a flame graph for Asterinas.

Just like what's proposed in #691, the first step is to launch a GDB server (`make gdb_server RELEASE=1`), attach to it (`make gdb_client`) and quit to leave the kernel running.

Then, use the following script to sample call stacks using GDB:

<details><summary>script to sample</summary>
<p>

```bash
#!/bin/bash

# Number of samples
nsamples=100
# Sleep time between samples (in seconds)
sleeptime=0.1

# Hostname or IP address of the machine running QEMU with GDB server
remote_host="localhost"
# Port number where QEMU GDB server is listening
remote_port="1234"

sleep 0.1

for x in $(seq 1 $nsamples)
do
	gdb -batch \
		-ex "set pagination 0" \
		-ex "file target/osdk/aster-nix/aster-nix-osdk-bin" \
		-ex "target remote $remote_host:$remote_port" \
		-ex "bt -frame-arguments presence -frame-info short-location" >> gdb_perf.log
	sleep $sleeptime
done
```

</p>
</details> 

After that, the text log is dumped to `gdb_perf.log`. Use a python script to generate folded stack traces for the flame graph:

<details><summary>Details</summary>
<p>

```python
import re

def process_stack_trace(file_path):
    with open(file_path, 'r') as file:
        lines = file.readlines()

    captures = []
    current_capture = []

    for line in lines:
        if not line.startswith('#'):
            continue
        if line.startswith('#0'):
            if current_capture:
                captures.append(current_capture)
                current_capture = []
        processed = line.strip()
        # remove all things between < and >, use bracket matching
        cut_generic = ''
        cnt = 0
        for c in processed:
            if c == '<':
                cnt += 1
            if cnt == 0:
                cut_generic += c
            if c == '>':
                cnt -= 1
        processed = cut_generic
        # remove all things like "::{impl#70}"
        processed = re.sub(r'::\{.*?\}', '', processed)
        # remove all "(...)"
        processed = processed.replace('(...)', '')
        # remove all "()"
        processed = processed.replace('()', '')
        # remove all things like "0xffffffff8819d0fb in"
        processed = re.sub(r'0x[0-9a-f]+ in', '', processed)

        # split by spaces, the first is number and the second is function name
        parts = [s for s in processed.split(' ') if s != '']

        current_capture.append(parts[1])

    if current_capture:
        captures.append(current_capture)

    folded = {} # { bt: value }

    for capture in captures:
        bt_from_butt = []
        for frame in reversed(capture):
            bt_from_butt.append(frame)
        folded_key = ';'.join(bt_from_butt)
        if folded_key in folded:
            folded[folded_key] += 1
        else:
            folded[folded_key] = 1

    with open('out.folded', 'w') as out_file:
        for key, v in folded.items():
            out_file.write(f"{key} {v}\n")

if __name__ == "__main__":
    process_stack_trace('gdb_perf.log')

```

</p>
</details> 

This script generates a file `out.folded`. Then the file is ready to be processed using [Flame graph](https://github.com/brendangregg/FlameGraph). Follow the guide to have an SVG.

TLDR:

```shell
./FlameGraph/flamegraph.pl ./asterinas/out.folded > kernel.svg
```

Here's an example on the unixbench spawn benchmark (on [c75a373](https://github.com/asterinas/asterinas/commit/c75a3732b9a6a7f0dbf11a839affaf2c126ecdc5)):

![kernel](https://github.com/asterinas/asterinas/assets/30975570/0ea5b99b-c769-4342-b17f-0d17e100ef8d)

([An interactive one](https://github.com/asterinas/asterinas/assets/30975570/0ea5b99b-c769-4342-b17f-0d17e100ef8d))

Here's also another example using #895 to optimize it:

![kernel_vmspacerw](https://github.com/asterinas/asterinas/assets/30975570/6e960bd7-ed26-49d0-91a5-1289673b6215)

([An interactive one](https://github.com/asterinas/asterinas/assets/30975570/6e960bd7-ed26-49d0-91a5-1289673b6215))

You can clearly see that when the bottleneck `read_val_from_user` is optimized, the performance boosts significantly and the bottleneck becomes the `ProcessBuilder`.
